requests/models.py

```diff
@@ -455,27 +453,32 @@ def prepare_body(self, data, files, json=None):
                     else:
                         content_type = 'application/x-www-form-urlencoded'
 
-            self.prepare_content_length(body)
-
             # Add content-type if it wasn't explicitly provided.
             if content_type and ('content-type' not in self.headers):
                 self.headers['Content-Type'] = content_type
 
+        self.prepare_content_length(body)
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
-            end_pos = body.tell()
-            self.headers['Content-Length'] = builtin_str(max(0, end_pos - curr_pos))
-            body.seek(curr_pos, 0)
-        elif body is not None:
-            l = super_len(body)
-            if l:
-                self.headers['Content-Length'] = builtin_str(l)
-        elif (self.method not in ('GET', 'HEAD')) and (self.headers.get('Content-Length') is None):
-            self.headers['Content-Length'] = '0'
+        """Prepares Content-Length header.
+        
+        If Transfer-Encoding header has been set, we want to make sure that Content-Length
+        header is not set. If setting Content-Length, we will strip Transfer-Encoding header.
+        """
+
+        if 'Transfer-Encoding' in self.headers:
+            if 'Content-Length' in self.headers:
+                del headers['Content-Length']
```

Comment:
@sigmavirus24 I added a test at the start of this issue to ensure that both headers are not set in the 0 length case, but I'll add another couple tests to ensure that it works correctly in the case that a user specifies a header that would be invalid.